### PR TITLE
Mempool

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/entities/altpoptx.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include "veriblock/entities/atv.hpp"
+#include "veriblock/entities/vtb.hpp"
+#include "veriblock/serde.hpp"
+#include "veriblock/slice.hpp"
+
+namespace altintegration {
+
+struct AltPopTx {
+  int32_t version{};
+
```
Comment (Warchant) [inline include/veriblock/entities/altpoptx.hpp:17]: where is vector of VbkBlocks?

File: include/veriblock/mempool.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    const std::vector<uint8_t>& bytes);
+
+struct MemPool {
+  using vbk_hash_t = decltype(VbkBlock::previousBlock);
+
+  ~MemPool() = default;
+  MemPool(Hash_Function /*function*/) {}
+
+  bool submitVTB(const std::vector<VTB>& vtb,
+                 const AltTree& tree,
```
Comment (Warchant) [inline include/veriblock/mempool.hpp:27]: why do you pass AltTree here? I thought it is needed only when you do `getPop`

File: include/veriblock/mempool.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                 ValidationState& state);
+  bool submitATV(const std::vector<ATV>& atv,
+                 const AltTree& tree,
+                 ValidationState& state);
```
Comment (Warchant) [inline include/veriblock/mempool.hpp:31]: same

File: include/veriblock/mempool.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  bool getPop(const AltBlock& current_block,
+              AltTree& tree,
+              AltPopTx* out_data,
```
Comment (Warchant) [inline include/veriblock/mempool.hpp:35]: why ptr? maybe use reference?

File: src/entities/atv.cpp [hunk 5]
```diff
@@ -7,10 +7,13 @@ ATV ATV::fromVbkEncoding(ReadStream& stream) {
...
   atv.containingBlock = VbkBlock::fromVbkEncoding(stream);
-  atv.context = readArrayOf<VbkBlock>(
-      stream, 0, MAX_CONTEXT_COUNT_ALT_PUBLICATION, [](ReadStream& stream) {
-        return VbkBlock::fromVbkEncoding(stream);
-      });
+  atv.context =
+      readArrayOf<VbkBlock>(stream,
+                            0,
+                            MAX_CONTEXT_COUNT_ALT_PUBLICATION,
+                            [](ReadStream& stream) -> VbkBlock {
```
Comment (Warchant) [inline src/entities/atv.cpp:14]: you can remove lambda and use `VbkBlock::fromVbkEncoding` instead

File: src/mempool.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#include "veriblock/mempool.hpp"
+
+#include "veriblock/stateless_validation.hpp"
+
+namespace altintegration {
+
+bool MemPool::submitATV(const std::vector<ATV>& atvs,
+                        const AltTree& tree,
```
Comment (Warchant) [inline src/mempool.cpp:8]: I think it is better to pass params in MemPool constructor rather than passing whole tree here

File: src/mempool.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    stored_atvs.push_back(atvs[i]);
+  }
+
+  return true;
+}
+
+bool MemPool::submitVTB(const std::vector<VTB>& vtbs,
+                        const AltTree& tree,
```
Comment (Warchant) [inline src/mempool.cpp:22]: same

File: src/mempool.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  static constexpr size_t vbk_prev_block_hash_size =
+      decltype(VbkBlock::previousBlock)::size();
+
+  if (!tree.setState(current_block.getHash(), state)) {
+    return state.Invalid("mempool-get-pop");
+  }
+
+  for (const auto& atv : stored_atvs) {
+    VbkBlock first_block =
+        atv.context.size() != 0 ? atv.context[0] : atv.containingBlock;
```
Comment (Warchant) [inline src/mempool.cpp:49]:
```suggestion
        !atv.context.empty()? atv.context[0] : atv.containingBlock;
```

**Comment by Warchant** [comment]
Also, pls rename `MemPool -> PopMemPool`

File: src/mempool.cpp (hunk 8 above)
Comment (blackyblack) [reply src/mempool.cpp:49]: Better to reverse comparison:
atv.context.empty()? atv.containingBlock : atv.context[0];

File: src/mempool.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  return true;
+}
+
+bool MemPool::submitVTB(const std::vector<VTB>& vtbs,
+                        const AltTree& tree,
+                        ValidationState& state) {
+  for (size_t i = 0; i < vtbs.size(); ++i) {
+    if (!checkVTB(
+            vtbs[i], state, tree.vbk().getParams(), tree.btc().getParams())) {
+      state.Invalid("mempool-submit-vtb", i);
```
Comment (blackyblack) [inline src/mempool.cpp:27]: Are you sure you want to continue with invalid data? I think you should return state.Invalid here.

Comment (Mr-Leshiy) [reply src/mempool.cpp]: fixed.

**Pull request merged**